### PR TITLE
Deployment Considerations documentation

### PR DESCRIPTION
**Pull request by gjoseph92**
This document tries to cover some of the infrastructure challenges _outside of Dask_ that people commonly run into when setting up serious (production, multi-tenant) Dask deployments. The goal here is to give a more realistic picture of what it takes to run a production-grade Dask deployment to people who might be thinking of setting one up.

This is spun out from https://github.com/dask/dask/pull/9912/files#r1096578227, and based loosely on @mrocklin's PyData NYC talk: https://www.youtube.com/watch?v=5hUkUj1VYW4.

cc @scharlottej13 @jacobtomlinson

### CODE REVIEW

File: docs/source/deployment-considerations.rst [hunk 1]
```diff
@@ -0,0 +1,91 @@
+Deployment Considerations
+=========================
+
+Thanks to the efforts of the open-source community, there are tools to deploy Dask :ref:`pretty much anywhere <deployment-options>`—if you can get computers to talk to each other, you can probably turn them into a Dask cluster.
+
+**However, getting Dask running is often not the last step, but the first step.** This document attempts to cover some of the things *outside of Dask* you may have to think about when managing a Dask deployment.
```
Comment (jacobtomlinson) [inline docs/source/deployment-considerations.rst:6]: I like how this is clearly setting the stage that these things are out of scope or on the periphery for Dask.

File: docs/source/deployment-considerations.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+Software environments can be particularly challenging when managing Dask for multiple users. Sometimes, using the same locked-down set of packages is sufficient for a team, but often individuals will need different packages, or different versions of the same package (including Dask itself!), to be productive. In these cases, you might end up giving end-users access to the build system (requiring them to build and maintain Docker images themselves, for example), or if that isn't allowable in your organization, creating custom infrastructure, or resorting to workarounds like the :class:`~distributed.diagnostics.plugin.PipInstall` or :class:`~distributed.diagnostics.plugin.UploadDirectory` plugins.
+
+Additional challenges can include getting local packages or scripts onto the cluster (and ensuring they're up to date), as well as packages installed from private Git or PyPI repos.
+
+
+Observability
```
Comment (jacobtomlinson) [inline docs/source/deployment-considerations.rst:22]: This is super important, but I feel like this is one of the last things folks think about. I would probably move this below other sections like cost and credentials.

File: docs/source/deployment-considerations.rst [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+Monitoring cost
+---------------
+It's good to be able to answer questions such as:
+
+- How much are we spending on Dask?
+- What are we spending it on? (machines, machines that should have been turned off, network egress that shouldn't have happened, etc.)
+- Who/what is responsible?
+
+Non-commercial deployment tools generally don't build in this sort of monitoring. Organizations that need it either end up building their own tools, or turning to commercial deployment offerings.
```
Comment (jacobtomlinson) [inline docs/source/deployment-considerations.rst:66]: Maybe soften this a bit.

```suggestion
Many deployment tools generally don't build in this sort of monitoring. Organizations that need it either end up building their own tools, or turning to commercial deployment offerings.
```

File: docs/source/deployment-considerations.rst [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+Managing networking
+-------------------
+The Dask client needs to be able to talk to the scheduler, which is potentially on a different system. Users like to be able to access the :doc:`dashboard <dashboard>` from a Web browser. The machines in the cluster need to be able to talk to each other. Typically, whatever :ref:`deployment system <deployment-options>` you use will handle this for you. Sometimes, though, there can be additional considerations around what type of networking to use for best performance. Networking also can have costs associated—cloud providers may charge fixed or usage-based rates for certain types of networking configurations, for example.
+
+You may also have other systems on restricted networks that workers need to access to read and write data, or call APIs. Connecting to those networks could add additional complexity.
+
+Some organizations may have additional network security policies, such as requiring all traffic to be encrypted. Dask supports this with :doc:`TLS <tls>`, which requires additional configuration, and managing certificates.
```
Comment (jacobtomlinson) [inline docs/source/deployment-considerations.rst:75]: Side note but Dask Cloud Provider turns this on by default. I wonder if we should do that in more deployment tooling and make it more of an opt out.

File: docs/source/deployment-considerations.rst [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+
+Storing local data off the local machine
+----------------------------------------
+if you're deploying Dask on a cluster, most data is probably already stored remotely, since a major reason for deploying Dask instead of :ref:`running locally <deployment-single-machine>` is to run workers closer to the data. However, it can be common to also have some smaller, auxiliary data files locally.
```
Comment (jacobtomlinson) [inline docs/source/deployment-considerations.rst:80]:
```suggestion
If you're deploying Dask on a cluster, most data is probably already stored remotely, since a major reason for deploying Dask instead of :ref:`running locally <deployment-single-machine>` is to run workers closer to the data. However, it can be common to also have some smaller, auxiliary data files locally.
```

File: docs/source/deployment-considerations.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+In that case, you may need somewhere to store those auxiliary files remotely, where workers can access them. Depending on your deployment system, there are many options, from network file systems to cloud object stores like S3. Regardless, this can be another piece of infrastructure to manage and secure.
+
+
+Note on managed Dask offerings
+------------------------------
+As shown, setting up and managing a mature Dask deployment, especially for team or production use, can involve a fair amount of complexity outside of Dask itself. Addressing these challenges is generally out of scope for the open-source Dask deployment tools, but there are commercial Dask deployment services that handle many of these considerations. In alphabetical order:
```
Comment (jacobtomlinson) [inline docs/source/deployment-considerations.rst:87]: I would maybe broaden this and also include projects like QHub.

```suggestion
As shown, setting up and managing a mature Dask deployment, especially for team or production use, can involve a fair amount of complexity outside of Dask itself. Addressing these challenges is generally out of scope for the open-source Dask deployment tools, but there are other projects as well as commercial Dask deployment services that handle many of these considerations. In alphabetical order:
```

**Review by jacobtomlinson — COMMENTED** [review]
I really like the way this has been laid out. It talks through a lot of important factors that advanced users will want to think about. Especially when rolling Dask out within orgs. As you say much of this is out of scope for core Dask.

I like how it starts with the challenges and then neatly directs users off to other projects and companies that solve those problems. This seems like a really great way to direct folks away to other projects and commercial offerings.

File: docs/source/deployment-considerations.rst (hunk 2 above)
Comment (gjoseph92) [reply docs/source/deployment-considerations.rst:22]: That was actually why I moved it up here; I know people don't usually think about it up front, but I wanted to make it more prominent since it's so important.

Also note that I mixed log retention in with metrics. Maybe those are worth splitting; I think log retention should be quite high (you're really not going to have a good time if you don't even keep logs around), but metrics usually come a bit later in your deployment journey.

File: docs/source/deployment-considerations.rst (hunk 3 above)
Comment (gjoseph92) [reply docs/source/deployment-considerations.rst:66]: I'd originally written that, but then couldn't think of any non-commercial tools that actually did have built-in capabilities for cost monitoring. Is there something I'm not thinking of?

Arguably Coiled doesn't even do what I've described here. Coiled can tell you how you spent your _Coiled_ bill, but for your AWS bill, you still have to look yourself in the cost explorer (though this is facilitated by tags Coiled adds to all your dask infrastructure).

**Comment by gjoseph92** [comment]
@jacobtomlinson I think I've addressed your comments!

**Review by jacobtomlinson — APPROVED** [review]
Thanks for writing this up @gjoseph92
